File: compatibility-tests/sdk-test-go/tests/ssm_test.go
package tests

import (
	"context"
	"testing"

	"floci-sdk-test-go/internal/testutil"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/ssm"
	ssmtypes "github.com/aws/aws-sdk-go-v2/service/ssm/types"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestSSM(t *testing.T) {
	ctx := context.Background()
	svc := testutil.SSMClient()
	name := "/go-test/param"
	value := "go-test-value"

	// Cleanup at end
	t.Cleanup(func() {
		svc.DeleteParameter(ctx, &ssm.DeleteParameterInput{Name: aws.String(name)})
	})

	t.Run("PutParameter", func(t *testing.T) {
		_, err := svc.PutParameter(ctx, &ssm.PutParameterInput{
			Name:      aws.String(name),
			Value:     aws.String(value),
			Type:      ssmtypes.ParameterTypeString,
			Overwrite: aws.Bool(true),
		})
		require.NoError(t, err)
	})

	t.Run("GetParameter", func(t *testing.T) {
		r, err := svc.GetParameter(ctx, &ssm.GetParameterInput{Name: aws.String(name)})
		require.NoError(t, err)
		assert.Equal(t, value, aws.ToString(r.Parameter.Value))
	})

	t.Run("GetParameters", func(t *testing.T) {
		r, err := svc.GetParameters(ctx, &ssm.GetParametersInput{Names: []string{name}})
		require.NoError(t, err)
		assert.Len(t, r.Parameters, 1)
	})

	t.Run("DescribeParameters", func(t *testing.T) {
		r, err := svc.DescribeParameters(ctx, &ssm.DescribeParametersInput{})
		require.NoError(t, err)
		assert.NotEmpty(t, r.Parameters)
	})

	t.Run("GetParametersByPath", func(t *testing.T) {
		r, err := svc.GetParametersByPath(ctx, &ssm.GetParametersByPathInput{
			Path: aws.String("/go-test"),
		})
		require.NoError(t, err)
		assert.NotEmpty(t, r.Parameters)
	})

	t.Run("DeleteParameter", func(t *testing.T) {
		_, err := svc.DeleteParameter(ctx, &ssm.DeleteParameterInput{Name: aws.String(name)})
		require.NoError(t, err)
	})
}

File: compatibility-tests/sdk-test-rust/tests/ssm_test.rs
mod common;

use aws_sdk_ssm::types::ParameterType;

#[tokio::test(flavor = "multi_thread")]
async fn test_ssm_put_parameter() {
    let ssm = common::ssm_client().await;
    let name = "/rust-test/param";
    let value = "rust-test-value";

    let _guard = common::CleanupGuard::new({
        let ssm = ssm.clone();
        async move {
            let _ = ssm.delete_parameter().name(name).send().await;
        }
    });

    let result = ssm
        .put_parameter()
        .name(name)
        .value(value)
        .r#type(ParameterType::String)
        .overwrite(true)
        .send()
        .await;

    assert!(result.is_ok(), "PutParameter failed: {:?}", result.err());
}

#[tokio::test(flavor = "multi_thread")]
async fn test_ssm_get_parameter() {
    let ssm = common::ssm_client().await;
    let name = "/rust-test/get-param";
    let value = "rust-get-value";

    let _guard = common::CleanupGuard::new({
        let ssm = ssm.clone();
        async move {
            let _ = ssm.delete_parameter().name(name).send().await;
        }
    });

    // Setup
    ssm.put_parameter()
        .name(name)
        .value(value)
        .r#type(ParameterType::String)
        .overwrite(true)
        .send()
        .await
        .expect("setup put");

    let result = ssm.get_parameter().name(name).send().await;
    assert!(result.is_ok(), "GetParameter failed: {:?}", result.err());

    let param = result.unwrap();
    assert_eq!(
        param.parameter().and_then(|p| p.value()).unwrap_or(""),
        value
    );
}

#[tokio::test(flavor = "multi_thread")]
async fn test_ssm_get_parameters() {
    let ssm = common::ssm_client().await;
    let name = "/rust-test/get-params";

    let _guard = common::CleanupGuard::new({
        let ssm = ssm.clone();
        async move {
            let _ = ssm.delete_parameter().name(name).send().await;
        }
    });

    // Setup
    ssm.put_parameter()
        .name(name)
        .value("test")
        .r#type(ParameterType::String)
        .overwrite(true)
        .send()
        .await
        .expect("setup");

    let result = ssm.get_parameters().names(name).send().await;
    assert!(result.is_ok(), "GetParameters failed: {:?}", result.err());
    assert_eq!(result.unwrap().parameters().len(), 1);
}

#[tokio::test(flavor = "multi_thread")]
async fn test_ssm_describe_parameters() {
    let ssm = common::ssm_client().await;
    let name = "/rust-test/describe";

    let _guard = common::CleanupGuard::new({
        let ssm = ssm.clone();
        async move {
            let _ = ssm.delete_parameter().name(name).send().await;
        }
    });

    // Setup
    ssm.put_parameter()
        .name(name)
        .value("test")
        .r#type(ParameterType::String)
        .overwrite(true)
        .send()
        .await
        .expect("setup");

    let result = ssm.describe_parameters().send().await;
    assert!(result.is_ok(), "DescribeParameters failed: {:?}", result.err());
    assert!(!result.unwrap().parameters().is_empty());
}

#[tokio::test(flavor = "multi_thread")]
async fn test_ssm_get_parameters_by_path() {
    let ssm = common::ssm_client().await;
    let name = "/rust-test/bypath/param";

    let _guard = common::CleanupGuard::new({
        let ssm = ssm.clone();
        async move {
            let _ = ssm.delete_parameter().name(name).send().await;
        }
    });

    // Setup
    ssm.put_parameter()
        .name(name)
        .value("test")
        .r#type(ParameterType::String)
        .overwrite(true)
        .send()
        .await
        .expect("setup");

    let result = ssm
        .get_parameters_by_path()
        .path("/rust-test/bypath")
        .send()
        .await;
    assert!(result.is_ok(), "GetParametersByPath failed: {:?}", result.err());
    assert!(!result.unwrap().parameters().is_empty());
}

#[tokio::test(flavor = "multi_thread")]
async fn test_ssm_delete_parameter() {
    let ssm = common::ssm_client().await;
    let name = "/rust-test/delete";

    // No cleanup guard needed - test is about deletion

    // Setup
    ssm.put_parameter()
        .name(name)
        .value("test")
        .r#type(ParameterType::String)
        .overwrite(true)
        .send()
        .await
        .expect("setup");

    let result = ssm.delete_parameter().name(name).send().await;
    assert!(result.is_ok(), "DeleteParameter failed: {:?}", result.err());
}
